aten/src/ATen/native/README.md

```diff
@@ -186,6 +189,49 @@ are appropriate for "core" Tensor operations (e.g., add, sub, etc.), but not for
 more complicated neural network layers (e.g., `conv2d`) and internal functions
 designed specifically for binding (e.g., `cudnn_convolution`).
 
+As we progress along our schema unification of the `func` schema with the JIT
+signatue schema, we must introduce features that allow us to increase compliance.
+One of these features are Tensor annotations. As of now we use naming conventions
+to indicate whether an argument of a function is going to be mutated and returned.
+
+There are two typical situations in which we mutate the memory of an argument in the Python
+frontend:
+a) For an inplace operations such as `self.abs_()`
+b) for a function with an output keyword argument such as `torch.abs(input, out=None)`.
+
+In order to provide implementations for these Python functions the legacy schema
+requires C++ implementations for three situations `abs(Tensor self)  -> Tensor`, 
+`abs_(Tensor self) -> Tensor` and `abs_out(Tensor out, Tensor self) -> Tensor`.
+
+Now, as we move towards the unification, we start to use a different syntax to represent
+this by using annotations. In the end we still translate to the legacy schema for the downstream
+consumers such as the C++ code generation, but this will soon change.
+
+If two Tensors carry the same annotation, they both *may* represent the same memory.
```

Comment:
there was a lot of discussion about *may* vs *must* and I'm admittedly still confused...

The situation discussed was `contiguous` which may return the same memory, if the Tensor is already contiguous.

If *may* is the correct answer, then an acceptable annotation is:
`contiguous(Tensor(a) self)-> Tensor(a)`

If *must* is the correct answer, then an acceptable annotation is:
`contiguous(Tensor(a) self)-> Tensor(a|b)` (i.e. it *must* be memory from `a` or memory from newly-introduced `b`).

Of course the second definition is also acceptable for *may*, but you'd never do that...

Reply:
It is indeed *may* not *must*. We don't use annotations rigorously yet as far as I can tell. This will become more apparent as we move along the unification and is a feature of this process.